File: Distribution-Grid-EV-CA/06_01_sum_charging_profiles.py
"""
06_01_sum_charging_profiles.py — Python port of ``06_01_sum_charging profiles.R``.

Produces per-feeder hourly load profiles for each year by expanding every
sampled charging session into a 24-hour vector and summing by feeder/type.
"""

from __future__ import annotations

import os

import numpy as np
import pandas as pd

from common import read_rds_like, save_rds_like

SESSION_FEEDER_PKL = (
    "data/result/charging session by feeder/"
    "session_feeder_new_bins_IOU_sample42_draw12_split6699_sample36.pkl"
)
OUT_CUM_PKL = (
    "data/result/charging profile by feeder/"
    "profile_feeder_cum_bins_IOU_sample42_draw12_split6699_sample36.pkl"
)

YEARS = range(2022, 2046)


def _year_profile(events: pd.DataFrame) -> pd.DataFrame:
    """Expand each event into 24 hourly rows and sum by (FeederID, charge_type, hour)."""
    n = len(events)
    if n == 0:
        return pd.DataFrame(columns=["FeederID", "charge_type", "hour", "load_feeder"])
    ev = events.loc[events.index.repeat(24)].reset_index(drop=True)
    ev["hour"] = np.tile(np.arange(24), n)
    ev["load"] = 0.0
    same_day = (ev["hour"] >= ev["start_hour"]) & (ev["hour"] <= ev["end_hour"])
    overnight = (ev["start_hour"] > ev["end_hour"]) & (
        (ev["hour"] >= ev["start_hour"]) | (ev["hour"] <= ev["end_hour"])
    )
    ev.loc[same_day, "load"] = ev.loc[same_day, "power"]
    ev.loc[overnight, "load"] = ev.loc[overnight, "power"]
    agg = ev.groupby(["FeederID", "charge_type", "hour"], as_index=False)["load"].sum()
    return agg.rename(columns={"load": "load_feeder"})


def main() -> None:
    os.makedirs(os.path.dirname(OUT_CUM_PKL), exist_ok=True)
    event_feeder = read_rds_like(SESSION_FEEDER_PKL)
    all_feeders = event_feeder["FeederID"].unique().tolist()

    # step 1: per-year new profiles (not cumulative)
    byyear_parts: list[pd.DataFrame] = []
    for y in YEARS:
        y_prof = _year_profile(event_feeder[event_feeder["year"] == y])
        y_prof["year"] = y
        byyear_parts.append(y_prof)
    by_year = pd.concat(byyear_parts, ignore_index=True)

    # step 2: cumulative up to year y, expanded to all feeders × {home, work, public} × 24h
    idx = pd.MultiIndex.from_product(
        [all_feeders, ["home", "work", "public"], range(24)],
        names=["FeederID", "charge_type", "hour"],
    )
    template = pd.DataFrame(index=idx).reset_index()

    cum_parts: list[pd.DataFrame] = []
    for y in YEARS:
        prev = by_year[by_year["year"] <= y]
        cum = (
            prev.groupby(["FeederID", "charge_type", "hour"], as_index=False)["load_feeder"]
            .sum()
            .rename(columns={"load_feeder": "load"})
        )
        merged = template.merge(cum, on=["FeederID", "charge_type", "hour"], how="left")
        merged["load"] = merged["load"].fillna(0.0)
        merged["year"] = y
        cum_parts.append(merged)

    cum_profile = pd.concat(cum_parts, ignore_index=True)
    save_rds_like(cum_profile, OUT_CUM_PKL)


if __name__ == "__main__":
    main()

File: Distribution-Grid-EV-CA/R_Yanning/06_01_sum_charging profiles.R
library(data.table)

event.feeder <- readRDS('data/result/charging session by feeder/session_feeder_new_bins_IOU_sample42_draw12_split6699_sample36.rds')
head(event.feeder)

length(unique(event.feeder[year==2022]$FeederID)) # 7926
length(unique(event.feeder[year==2023]$FeederID)) # 7874
length(unique(event.feeder$FeederID)) # 7958/8032

## add up new charging profile separately by year (by feeder by type)
feeder.prof.new.byyear <- data.table()
for (y in 2022:2045) {
  event.feeder.temp <- event.feeder[year==y]
  # 24h charging profile for each event
  feeder.prof.temp <- event.feeder.temp[rep(1:nrow(event.feeder.temp),each=24),]
  feeder.prof.temp[,hour:=rep(0:23,times=nrow(event.feeder.temp))]
  feeder.prof.temp[,load:=0]
  # same day
  feeder.prof.temp[hour>=start_hour & hour<=end_hour,load:=power] 
  # charge till tomorrow, add back to the beginning of today
  feeder.prof.temp[start_hour>end_hour & (hour>=start_hour|hour<=end_hour),load:=power] 
  feeder.prof.temp <- feeder.prof.temp[,.(load_feeder=sum(load)),by=.(FeederID,charge_type,hour)]
  feeder.prof.temp[,year:=y]
  
  feeder.prof.new.byyear <- rbind(feeder.prof.new.byyear,feeder.prof.temp)
}

head(feeder.prof.new.byyear)

## add up charging profile cumulatively by year (by feeder by type)
feeder.prof.cum.byyear <- data.table()
# for now, consider only the feeders that have EV charging load
# if needed, can be changed into all valid feeders from the feeder-block mapping table
all.feeder <- unique(event.feeder$FeederID)

for (y in 2022:2045){
  # make space for all feeders (EV load does not increase in every feeder each year)
  prof.temp.allfeeder <- data.table(FeederID=rep(all.feeder,each=3),
                                    charge_type=rep(c('home','work','public'),times=length(all.feeder)))
  prof.temp.allfeeder <- prof.temp.allfeeder[rep(1:nrow(prof.temp.allfeeder),each=24),]
  prof.temp.allfeeder[,hour:=rep(0:23,times=length(all.feeder)*3)]
  # add up the existing profile for all years so far
  feeder.prof.new.temp <- feeder.prof.new.byyear[year<=y]
  feeder.prof.cum.temp <- feeder.prof.new.temp[,.(load=sum(load_feeder)),by=.(FeederID,charge_type,hour)]
  # merge it into all feeders
  prof.temp.allfeeder <- feeder.prof.cum.temp[prof.temp.allfeeder,on=.(FeederID,charge_type,hour)]
  # the feeders where EV load haven't appeared yet should be set to 0
  prof.temp.allfeeder[is.na(load),load:=0]
  
  prof.temp.allfeeder[,year:=y]
  feeder.prof.cum.byyear <- rbind(feeder.prof.cum.byyear,prof.temp.allfeeder)
}

head(feeder.prof.cum.byyear)

saveRDS(feeder.prof.cum.byyear,'data/result/charging profile by feeder/profile_feeder_cum_bins_IOU_sample42_draw12_split6699_sample36.rds')
